### PR TITLE
43 retrain model q1 2025

### LINKED ISSUE #43: Retrain model with Q1 2025

### PR DESCRIPTION
**Pull request by sofiapinto**
---

# Description

In this PR we:
- retrain the model with Q1 2025 data
- move cross validation & hyperparameter code from notebooks to scripts
- make some arguments more explicit

Fixes #43 

# Instructions for Reviewer

Could you please review:
- Changes to the scripts in the following order:
   - asf_hp_cost_estimator_model/config/base.yaml
   - asf_hp_cost_estimator_model/utils/model_evaluation_utils.py (this is code you've seen but refactored)
   - asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py this is code you've seen but refactored)
   - asf_hp_cost_estimator_model/pipeline/model_evaluation/cross_validation.py (this is code you've seen but refactored)
   - asf_hp_cost_estimator_model/pipeline/model_training/fit_cost_prediction_intervals.py
   - asf_hp_cost_estimator_model/pipeline/data_processing/process_installations_data.py (you can even skip this one if you don't have time, i just made some arguments more explicit)

and please let me know if you find any logic error in the scripts.

Thank you so much in advance! :) 

# Checklist:

- [x] I have checked the code runs
- [ ] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [x] I have explained this PR above
- [ ] I have requested a code review

### CODE REVIEW

File: asf_hp_cost_estimator_model/config/base.yaml [hunk 1]
```diff
@@ -31,8 +31,8 @@ lad_to_region_census_2011_filename: "Local_Authority_District_to_Region_(April_2
 
 ### MCS EPC DATA ###
 
-# MCS-EPC data up to Q3 2024 (date is the date of processing)
-mcs_epc_filename_date: "250310"
+# MCS-EPC data up to Q1 2025 (date is the date of processing)
+mcs_epc_filename_date: "250609"
```
Comment (crispy-wonton) [inline asf_hp_cost_estimator_model/config/base.yaml:35]: As before, this could be changed to use the latest file and not hard-coded but I think if you do that it should be in a separate PR :)

File: asf_hp_cost_estimator_model/utils/model_evaluation_utils.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    lower_below_8k = y_pred_lower < 8000
+    logging.info(
+        f"Percentage of instances where lower bound is below 8k: {np.mean(lower_below_8k) * 100:.2f}%"
+    )
+
+    # average width of intervals for which lower bound is below 8k (but not the upper end) after replacing lower bound with 8k
+    lower_below_8k_upper_above = lower_below_8k & (y_pred_upper > 8000)
+    adjusted_lower_bound = np.where(lower_below_8k_upper_above, 8000, y_pred_lower)
+    adjusted_width = np.mean(
+        y_pred_upper - adjusted_lower_bound, where=lower_below_8k_upper_above
```
Comment (crispy-wonton) [inline asf_hp_cost_estimator_model/utils/model_evaluation_utils.py:114]: did not know you can use where in the `np.mean()` - cool!

File: asf_hp_cost_estimator_model/utils/model_evaluation_utils.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    )
+
+    # largest interval after readjusting lower bound
+    largest_interval = np.max(y_pred_upper - adjusted_lower_bound)
+    logging.info(
+        f"Largest interval after adjusting lower bound to 8k: {largest_interval:.2f}"
+    )
+
+    logging.info(
+        "**Although not very meaningful, calculating mean absolute error for both bounds**"
```
Comment (crispy-wonton) [inline asf_hp_cost_estimator_model/utils/model_evaluation_utils.py:133]: LOL. Yeh not sure how to interpret those numbers - any thoughts?

File: asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+Usage:
+python asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py --lower_quantile 0.1 --upper_quantile 0.9
+"""
+
+# package imports
+import numpy as np
+from sklearn.model_selection import train_test_split
+from sklearn.metrics import make_scorer, mean_pinball_loss
+from sklearn.experimental import enable_halving_search_cv  # noqa
```
Comment (crispy-wonton) [inline asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py:14]: I googled 'noqa' and i got 'Nordic quality assurance' 😆 

I can see that it's here so the linter ignores it. Didn't know you could do this - useful!

File: asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+if __name__ == "__main__":
+    args = argparse_setup()
+    lower_quantile = args.lower_quantile
+    upper_quantile = args.upper_quantile
+
+    # Load and process data
+    mcs_epc_data = get_enhanced_installations_data()
+    cpi_05_3_df = get_cpi_data()
+    cpi_quarterly_df = get_df_quarterly_cpi_with_adjustment_factors(
+        ref_year=config["cpi_data"]["cpi_reference_year"],
```
Comment (crispy-wonton) [inline asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py:73]: I'm not sure what the plan is for updating the model in the future, but thinking ahead, maybe we don't want this reference year to be hard coded? We could set it so that it takes the latest full year?

Not for this PR but suggestion for future.

File: asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+    # Split the data into training and test sets
+    x_train, x_test, y_train, y_test = train_test_split(
+        X, y, test_size=0.2, random_state=config["random_state"]
+    )
+
+    # Define the parameter grid for hyperparameter tuning
+    param_grid = dict(
+        learning_rate=[0.01, 0.05, 0.1, 0.2],
+        max_depth=[3, 5, 10, 20],
+        min_samples_leaf=[1, 5, 10, 100, 10000],
```
Comment (crispy-wonton) [inline asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py]:
```suggestion
        min_samples_leaf=[1, 5, 10, 100, 1000],
```
Is this meant to be 1000?

File: asf_hp_cost_estimator_model/pipeline/model_evaluation/cross_validation.py [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+    numeric_features = config["numeric_features"]
+    original_categorical_features = config["categorical_features_to_dummify"]
+    categorical_features = config["categorical_features"]
+    target_feature = config["target_feature"]
+
+    # Preparing data for modelling
+    X = model_data[numeric_features + categorical_features]
+    y = model_data[target_feature].values.ravel()
+
+    kf = KFold(n_splits=5, shuffle=True, random_state=42)
```
Comment (crispy-wonton) [inline asf_hp_cost_estimator_model/pipeline/model_evaluation/cross_validation.py]: Not important but just noting that elsewhere you've used the random state from the `config` file - in case you wanted to do it here too

File: asf_hp_cost_estimator_model/pipeline/model_evaluation/cross_validation.py [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+
+    # Lists to store metrics for each fold
+    list_train_mean_minball_loss_lower_perc = []
+    list_train_mean_minball_loss_upper_perc = []
```
Comment (crispy-wonton) [inline asf_hp_cost_estimator_model/pipeline/model_evaluation/cross_validation.py]:
```suggestion
    list_train_mean_pinball_loss_lower_perc = []
    list_train_mean_pinball_loss_upper_perc = []
```
hehehe just noticing this - it's in quite a few places so maybe not worth fixing!

File: asf_hp_cost_estimator_model/pipeline/model_evaluation/cross_validation.py [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+
+        model_lower = GradientBoostingRegressor(
+            loss="quantile",
+            alpha=lower_quantile,
+            n_estimators=config["n_estimators"],
+            min_samples_leaf=config["min_samples_leaf"],
+            min_samples_split=config["min_samples_split"],
+            random_state=config["random_state"],
+            learning_rate=config["learning_rate"],
+            max_depth=config["max_depth"],
```
Comment (crispy-wonton) [inline asf_hp_cost_estimator_model/pipeline/model_evaluation/cross_validation.py]: I assume these are the hyperparameters selected from your grid search.

File: asf_hp_cost_estimator_model/pipeline/model_evaluation/cross_validation.py [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+            list_coverage=list_test_coverage,
+            list_avg_width=list_test_avg_width,
+            y=y_test,
+            y_pred_upper=y_pred_upper_test,
+            y_pred_lower=y_pred_lower_test,
+            alpha_lower=lower_quantile,
+            alpha_upper=upper_quantile,
+        )
+
+    logging.info("----- MODEL EVALUATION RESULTS ON TRAINING SET -----")
```
Comment (crispy-wonton) [inline asf_hp_cost_estimator_model/pipeline/model_evaluation/cross_validation.py:198]: You could print out the lists for each metric as well as the mean average to see how variable the results are

File: asf_hp_cost_estimator_model/pipeline/model_training/fit_cost_prediction_intervals.py [hunk 11]
```diff
@@ -55,6 +58,26 @@ def argparse_setup():
     return parser.parse_args()
 
 
+def create_df_with_predictions(
+    y: np.array, y_pred_lower: np.array, y_pred_upper: np.array
+) -> pd.DataFrame:
+    """
+    Creates a DataFrame with predictions for the lower and upper quantiles.
+    Args:
+        X (np.array): Feature matrix.
```
Comment (crispy-wonton) [inline asf_hp_cost_estimator_model/pipeline/model_training/fit_cost_prediction_intervals.py:67]: Looks like `X` is not used or defined

**Review by crispy-wonton — COMMENTED** [review]
Hey @sofiapinto 
I couldnt find any bugs in the code. I did leave some minor comments for cleaning up some small typos. I left some questions and some suggestions for future automation to remove additional hard-coding.

One question I have for you - is the problem of the winsorisation + cost adjustment solved now? I've lost track a bit of where that happens. Did you decide on a final order to winsorise and then adjust for inflation?

File: asf_hp_cost_estimator_model/config/base.yaml (hunk 1 above)
Comment (sofiapinto) [reply asf_hp_cost_estimator_model/config/base.yaml:35]: Yes - having this hardcoded is a conscious decision for now! And it will be in the README.

File: asf_hp_cost_estimator_model/utils/model_evaluation_utils.py (hunk 3 above)
Comment (sofiapinto) [reply asf_hp_cost_estimator_model/utils/model_evaluation_utils.py:133]: something like... How far each bound is from the actual value? But yeah... not that useful

File: asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py (hunk 4 above)
Comment (sofiapinto) [reply asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py:14]: Found it online in a tutorial I think, also haven't used it before.

File: asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py (hunk 5 above)
Comment (sofiapinto) [reply asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py:73]: Yep, this is also consciously hardcoded for now.

Comment (sofiapinto) [reply asf_hp_cost_estimator_model/pipeline/hyperparameter_tuning/tune_hyperparameters.py:73]: Will be in the README

File: asf_hp_cost_estimator_model/pipeline/model_evaluation/cross_validation.py (hunk 9 above)
Comment (sofiapinto) [reply asf_hp_cost_estimator_model/pipeline/model_evaluation/cross_validation.py]: correct!

File: asf_hp_cost_estimator_model/pipeline/model_evaluation/cross_validation.py (hunk 8 above)
Comment (sofiapinto) [reply asf_hp_cost_estimator_model/pipeline/model_evaluation/cross_validation.py]: ahahah

**Comment by sofiapinto** [comment]
Thank you for the prompt and useful review @crispy-wonton ❤️ as always! yes the winsorisation + cost adjustment problem was solved in a separate PR.
I went with your suggestion since we're not using the adjusted cost as our target feature.
